janus_server/src/aggregator.rs

```diff
@@ -439,11 +442,163 @@ where
+                            let mut report_agg = report_aggregations.next().ok_or_else(|| {
+                                warn!(?task_id, ?job_id, nonce = %transition.nonce, "Leader sent unexpected or out-of-order transitions");
+                                datastore::Error::User(Box::new(Error::UnrecognizedMessage(
+                                    "leader sent unexpected or out-of-order transitions",
+                                    task_id,
+                                )))
+                            })?;
+                            if report_agg.nonce != transition.nonce {
+                                // TODO(brandon): assuming we aren't going to fail due to
+                                // out-of-order/unexpected transitions, any report aggregations that
+                                // get skipped here are skipped because the leader omitted them
+                                // because something failed on the leader's side. We move to INVALID
+                                // state to ensure we don't get later updates, but is that the
+                                // desired behavior?
+                                report_agg.state = ReportAggregationState::Invalid;
+                                tx.update_report_aggregation(&report_agg).await?;
+                                continue;
+                            }
+                            break report_agg;
+                        };
+                        let prep_step =
+                            match report_aggregation.state {
+                                ReportAggregationState::Waiting(prep_step) => prep_step,
+                                _ => {
+                                    warn!(?task_id, ?job_id, nonce = %transition.nonce, "Leader sent transition for non-WAITING report aggregation");
+                                    return Err(datastore::Error::User(Box::new(
+                                        Error::UnrecognizedMessage(
+                                            "leader sent transition for non-WAITING report aggregation",
+                                            task_id,
+                                        ),
+                                    )));
+                                },
+                            };
+
+                        // Parse preparation message out of transition received from leader.
+                        let prep_msg = match &transition.trans_data {
+                            TransitionTypeSpecificData::Continued { payload } => {
+                                A::PrepareMessage::decode_with_param(
+                                    &prep_step,
+                                    &mut Cursor::new(payload),
+                                )?
+                            }
+                            _ => {
+                                // TODO(brandon): should this be reflected in the response?
```

Comment:
Yes, the helper state machine says that if the leader did not continue, then the helper transitions to INVALID. Then, there's a sentence that says "If any state transition results in INVALID, the helper MUST abort with error 'unrecognizedMessage'".

Reply:
Ah, this presents a conundrum. In this and other cases, we want to both record some state in the database, and return a "400 Bad Request" with problem details as a response. However, when we return `Err(Error::UnrecognizedMessage(...))`, `run_tx_once()` exits early, and the transaction rolls back. I think our transaction closures will either need to return a Result with a special error type, so that we can commit on certain errors and roll back on others, or a nested Result (which could be an ergonomic pain without `#![feature(try_blocks)]`).

Reply:
I updated the semantics to abort as expected, thanks!

I think it's currently underspecified what to do if the entire request fails with `unrecognizedMessage` (or any other error): the leader has no way to know _which_ condition caused the failure, and therefore has no idea what state the helper might be in. In general, I'd be pretty surprised if a "bad request" (`unrecognizedMessage`) response caused state changes. I think the semantics we actually want are to not change state at all if the entire request is bad. TBF, that isn't what's written in the spec (i.e. in this case it will transition to INVALID and then abort), but I think we probably want to standardize on "aborting causes no state changes" -- which we'd then interpret as the helper realizing there's an INVALID transition, rolling back (undoing the invalid transition), and then responding with an error.